keps/sig-api-machinery/20190425-structural-openapi.md

```diff
@@ -0,0 +1,812 @@
+        XEmbeddedResource bool
+        
+        // x-kubernetes-int-or-string specifies that this value is 
+        // either an integer or a string. If this is true, an empty
+        // type is allowed and type as child of anyOf is permitted
+        // if following one of the following patterns:
+        //
+        // 1) anyOf:
+        //    - type: integer
+        //    - type: string
+        // 2) allOf:
+        //    - anyOf:
+        //      - type: integer
+        //      - type: string
+        //    - <NestedValueValidation>
+        XIntOrString bool
+}
+
+type Union struct {
+	// discriminator is the name of the discriminator field.
+	// +optional
+	Discriminator string
+	// discriminatedFields lists the union field names with their discriminator value.
+	// This must be undefined if discriminator is empty.
+	// +optional
+	DiscriminatedFields map[string]string
+	// fields lists the union field names.
+	// This must be undefined if discriminator is non-empty.
+	// +optional
+	Fields              []string
+}
+
+type Generic struct {
+	// type specifies the type of a value.
+	// It can be object, array, number, integer, boolean, string.
+	// It is optional only if x-kubernetes-preserve-unknown-fields 
+	// or x-kubernetes-int-or-string is true.
+	Type        string
+	AdditionalProperties *StructuralOrBool
+	
+	Description string
+	Title       string
+        Nullable    bool
+	Default     *JSON
+        ReadOnly    bool
+}
+
+type ValueValidation struct {
+	Format               string
+	Maximum              *float64
+        ExclusiveMaximum     bool
+        Minimum              *float64
+        ExclusiveMinimum     bool
+        MaxLength            *int64
+        MinLength            *int64
+        Pattern              string
+        MaxItems             *int64
+        MinItems             *int64
+        UniqueItems          bool
+        MultipleOf           *float64
+        Enum                 []JSON
+        MaxProperties        *int64
+        MinProperties        *int64
+        Required             []string
+        AllOf                []NestedValueValidation
+        OneOf                []NestedValueValidation
+        AnyOf                []NestedValueValidation
+        Not                  *NestedValueValidation
+}
+
+type NestedValueValidation struct {
+	ValueValidation
+
+	Items      *NestedValueValidation
+        Properties map[string]NestedValueValidation
+        
+        // Anything set in the following will make the scheme 
+        // non-structural, with the exception of these two patterns if 
+        // x-kubernetes-int-or-string is true:
+        //
+        // 1) anyOf:
+        //    - type: integer
+        //    - type: string
+        // 2) allOf:
+        //    - anyOf:
+        //      - type: integer
+        //      - type: string
+        //    - <NestedValueValidation>
+        ForbiddenGenerics Generic `"json:inline"`
+        ForbiddenExtensions Extensions `"json:inline"`
```

Comment:
can we drop the json annotations to make it clearer that this isn't used for serialization yet?